mono/eglib/glib.h

```diff
@@ -54,6 +54,248 @@
 #define G_END_DECLS
 #endif
 
+#ifdef __cplusplus
+
+#define g_cast_t monoeg_g_cast_t // in case not inlined (see eglib-remap.h)
+
+// Mostly-internal type to support g_cast(), but also can be used in containers.
+template <typename TFrom> // TFrom is almost always void*
+struct g_cast_t
+{
+private:
+	TFrom x;
+public:
+	explicit g_cast_t (TFrom y) : x(y)
+	{
+	}
+
+	g_cast_t () // Not really intended but has a user.
+	{
+	}
+
+	g_cast_t (const g_cast_t& y) : x(y.x)
+	{
+	}
+
+	template <typename TTo>
+	operator TTo()
+	{
+		return (TTo)x;
+	}
+
+	g_cast_t& operator= (TFrom y)
+	{
+		x = y;
+		return *this;
+	}
+
+	template <typename T>
+	bool operator != (T y) // e.g. compare to NULL
+	{
+		return x != (TFrom)y;
+	}
+
+	template <typename T>
+	bool operator == (T y) // e.g. compare to NULL
+	{
+		return x == (TFrom)y;
+	}
+
+	template <typename T>
+	bool operator < (T y)
+	{
+		return x < (TFrom)y;
+	}
+
+	template <typename T>
+	bool operator <= (T y)
+	{
+		return x <= (TFrom)y;
+	}
+
+	template <typename T>
+	bool operator > (T y)
+	{
+		return x > (TFrom)y;
+	}
+
+	template <typename T>
+	bool operator >= (T y)
+	{
+		return x >= (TFrom)y;
+	}
+};
+
+// Member function compares g_cast_t to other, global compares other to g_cast_t,
+// i.e. which parameter is first vs. second.
+template <typename T1, typename T2>
+inline bool operator == (T1 x, g_cast_t<T2> y) // e.g. compare to NULL
```

Comment:
First: `x` is `private`, so these all need to be `friend`s.

Second: this is a really really bad idea.

You're essentially throwing away all type safety (because `T1` and `T2` may have nothing to do with each other) so this will mask a ton of unintended bugs:

```cplusplus
int* f (void)
{
  return 0;
}

int
main ()
{
  g_cast_t<int*> z = g_cast ((int*)0);
  std::printf ("%d\n", f () == z);
  std::printf ("%d\n", f == z);
}
```

So in this case I accidentally wrote `f` instead of `f()` on the last line, and the answer changes from always true to always false without a warning (even with `-Wall -Wextra`).

Reply:
‘Private’ was a very late addition. Probably a mistake. Let’s just remove it.

Reply:
Type safety is not a goal here yet. Limiting line damage and making the code valid C & C++ & with same meaning.

But if you want TFrom only void* (or MonoObject*? Or any pointer?) and TTo only a pointer, then yes, agreed, I can try.

Reply:
But this also can help with function pointers fyi. But maybe fix those otherwise.